gitlab/client.py

```diff
@@ -1174,6 +1174,8 @@ def __init__(
         # Preserve kwargs for subsequent queries
         self._kwargs = kwargs.copy()
 
+        self._prev_page_objects = []
```

Comment:
Are we sure it's enough to just compare against the previous page? Or is there a chance that an item from one of the earlier pages is repeated (e.g. by using a certain sort parameter that ends up returning an old item again)?

Reply:
This duplication occurs when the sorting is done by ID and new objects are created because the new objects now exist on pages previously retrieved and the pagination requests have no context (i.e. no dedicated database curser for that query).  It's like in the UI looking at page 2 of something, then creating a new object in another tab, going back and refreshing page 2... everything shifted and stuff from page 1 (which you already retreived) is now on page 2.

I believe this duplication is independent of sort order but more pronounced with the default ID sorting because any new objects will result in duplication.

Let's think about any other kind of sorting... for example use alphabetical sorting.
In this case new objects which would appear on pages higher than your current page would not result in duplication, but objects which would appear on previous pages would result in duplication.
You're listing projects, you've got all the A's, B's, and are in the middle of listing projects starting with C's.  A new project called `Zoo` is created.  You will not see duplication, but if a new project called `Apples` is created the shift will happen and you'll get a duplicated project starting with C that you got on the previous page.

I do not believe it's necessary to store any extra items other than the ones from the previous page.

Reply:
I almost hate to mention it but there may be a case where the previous page is not enough, but it doesn't have to do with other sort parameters.

If between pages, more items are created on previous pages than the `per_page` number, then it would be possible for duplicates to occur.

The rarity of this occurring depends heavily on how the user code is implemented.

```python
# This code has higher possibility of 100 projects being created between pagination calls
for project in gl.projects.list(iterator=True):
  something_that_takes_a_long_time(project)

# This code has lower possibility of 100 projects being created between pagination calls
projects = gl.projects.list(get_all=True)
for project in projects:
  something_that_takes_a_long_time(project)
```

If we're worried about `OutOfMemory` errors as @max-wittig mentioned, the entire thing could be refactored to work off of sets of integers (objects IDs) instead of list of objects.